backend/groth16/bn254/solidity.go

```diff
@@ -490,18 +632,77 @@ contract Verifier {
     /// proof was successfully verified.
     /// @param proof the points (A, B, C) in EIP-197 format matching the output
     /// of compressProof.
+    {{- if gt $numCommitments 0 }}
+    /// @param commitments the Pedersen commitments from the proof.
+    /// @param commitmentPok the proof of knowledge for the Pedersen commitments.
+    {{- end }}
     /// @param input the public input field elements in the scalar field Fr.
     /// Elements must be reduced.
     function verifyProof(
         uint256[8] calldata proof,
-        uint256[{{$numPublic}}] calldata input
+        {{- if gt $numCommitments 0}}
+        uint256[{{mul 2 $numCommitments}}] calldata commitments,
+        uint256[2] calldata commitmentPok,
+        {{- end }}
+        uint256[{{$numWitness}}] calldata input
     ) public view {
+        {{- if eq $numCommitments 0 }}
         (uint256 x, uint256 y) = publicInputMSM(input);
+        {{- else }}
+        // HashToField
+        uint256[{{$numCommitments}}] memory publicCommitments;
+        {{- range $i := intRange $numCommitments }}
+            publicCommitments[{{$i}}] = uint256(
+                sha256(
+                    abi.encodePacked(
+                        commitments[{{mul $i 2}}],
+                        commitments[{{sum (mul $i 2) 1}}]
+                        {{- $pcIndex := index $PublicAndCommitmentCommitted $i }}
+                        {{- range $j := intRange (len $pcIndex) }}
+                        {{- $l := index $pcIndex $j }}
+                        ,input[{{sub $l 1}}]
+                        {{- end }}
+                    )
+                )
+            ) % R;
+        {{- end }}
+
+        // Verify pedersen commitments
+        bool success;
+        assembly ("memory-safe") {
+            let f := mload(0x40)
+
+            calldatacopy(f, commitments, 0x40) // Copy Commitments
+            mstore(add(f, 0x40), PEDERSEN_G_X_1)
+            mstore(add(f, 0x60), PEDERSEN_G_X_0)
+            mstore(add(f, 0x80), PEDERSEN_G_Y_1)
+            mstore(add(f, 0xa0), PEDERSEN_G_Y_0)
+            calldatacopy(add(f, 0xc0), commitmentPok, 0x40)
+            mstore(add(f, 0x100), PEDERSEN_GROOTSIGMANEG_X_1)
+            mstore(add(f, 0x120), PEDERSEN_GROOTSIGMANEG_X_0)
+            mstore(add(f, 0x140), PEDERSEN_GROOTSIGMANEG_Y_1)
+            mstore(add(f, 0x160), PEDERSEN_GROOTSIGMANEG_Y_0)
+
+            success := staticcall(gas(), PRECOMPILE_VERIFY, f, 0x180, f, 0x20)
```

Comment:
is it possible to merge this with the next batch of pairings?

Reply:
I am not sure the soundness if we combine two pairings. If pairings are correct then single pairing also works but I am not sure the other part

Reply:
I wonder if this could be done with a random linear combination? Would be a huge gas saving.

Reply:
I haven't written it out, but could be doable. And indeed, it would be a huge saving both for the in-circuit recursive verifier and Solidity verifier. I'll try and ask.

Reply:
> I haven't written it out, but could be doable. And indeed, it would be a huge saving both for the in-circuit recursive verifier and Solidity verifier. I'll try and ask.

Can confirm merging indeed work as I did it, but no clue security wise if this is sound...

Reply:
For 1 commitment, this saves approximately 70K gas

Reply:
Hmm, I'd assume if the coefficient for random linear combination comes from Fiat-Shamir, then I would by default assume it is sound. And we would need it anyway as in Solidity and in circuit we do not have access to RNG.

Reply:
> Hmm, I'd assume if the coefficient for random linear combination comes from Fiat-Shamir, then I would by default assume it is sound. And we would need it anyway as in Solidity and in circuit we do not have access to RNG.

Doing the shamir trick on the proof commitment G1 point and proof commitment proof of knowledge G1 point is fine (or also we also need to do it with A/C and H^alpha from the groth16 proof)?

I'm currently doing:
$`r = hash(PC, PCPPOK)`$
$`e(A, B) . e(L, -H^{\gamma}) . e(C, -H^{\delta}) . e(PC^r, g) . e(PCPOK^r, g_{-1/\sigma})) = 1_GT`$

Every extra multiplication for the random linear combination is costing slightly more than 6K gas

Reply:
For reference https://xn--2-umb.com/23/groth16-batch/#poor-man-s-snarkpack